File: modules/stego-core/src/main/java/com/shterneregen/securelan/stego/StegoServices.java
package com.shterneregen.securelan.stego;

import com.shterneregen.securelan.crypto.CryptoServices;
import com.shterneregen.securelan.stego.service.SteganographyService;
import com.shterneregen.securelan.stego.service.impl.BmpSteganographyService;

import java.util.Objects;

public final class StegoServices {
    private final SteganographyService steganographyService;

    private StegoServices(CryptoServices cryptoServices) {
        Objects.requireNonNull(cryptoServices, "cryptoServices");
        this.steganographyService = new BmpSteganographyService(cryptoServices.passwordFileCryptoWorkflow());
    }

    public static StegoServices createDefault() {
        return new StegoServices(CryptoServices.createDefault());
    }

    public static StegoServices create(CryptoServices cryptoServices) {
        return new StegoServices(cryptoServices);
    }

    public SteganographyService steganographyService() {
        return steganographyService;
    }
}

File: modules/stego-core/src/main/kotlin/com/shterneregen/securelan/stego/StegoServices.kt
package com.shterneregen.securelan.stego

import com.shterneregen.securelan.crypto.CryptoServices
import com.shterneregen.securelan.stego.service.SteganographyService
import com.shterneregen.securelan.stego.service.impl.BmpSteganographyService
import com.shterneregen.securelan.stego.service.StegoInspectionService
import com.shterneregen.securelan.stego.service.impl.BmpStegoInspectionService
import java.util.Objects

class StegoServices private constructor(cryptoServices: CryptoServices) {
    private val steganographyService: SteganographyService

    init {
        Objects.requireNonNull(cryptoServices, "cryptoServices")
        steganographyService = BmpSteganographyService(cryptoServices.passwordFileCryptoWorkflow())
    }

    fun steganographyService(): SteganographyService = steganographyService

    fun inspectionService(): StegoInspectionService = BmpStegoInspectionService()

    companion object {
        @JvmStatic
        fun createDefault(): StegoServices = StegoServices(CryptoServices.createDefault())

        @JvmStatic
        fun create(cryptoServices: CryptoServices): StegoServices = StegoServices(cryptoServices)
    }
}
